contracts/axelarnet-gateway/src/contract/execute.rs

```diff
@@ -88,31 +98,55 @@ pub fn call_contract(
         ),
     )
     .change_context(Error::InvalidCrossChainId)?;
-    let source_address = Address::from_str(sender.as_str())
-        .change_context(Error::InvalidSourceAddress(sender.clone()))?;
+    let source_address = Address::from_str(info.sender.as_str())
+        .change_context(Error::InvalidSourceAddress(info.sender.clone()))?;
     let msg = call_contract.to_message(id, source_address);
 
     state::save_unique_routable_msg(storage, &msg.cc_id, &msg)
         .inspect_err(|err| panic_if_already_exists(err, &msg.cc_id))
         .change_context(Error::SaveRoutableMessage)?;
 
-    Ok(
-        route_to_router(storage, &Router::new(router), vec![msg.clone()])?.add_event(
-            AxelarnetGatewayEvent::ContractCalled {
-                msg,
+    let res = match info.token().change_context(Error::InvalidToken)? {
+        None => {
+            let event = AxelarnetGatewayEvent::ContractCalled {
+                msg: msg.clone(),
                 payload: call_contract.payload,
-            }
-            .into(),
-        ),
-    )
+                token: None,
+            };
+
+            route_to_router(storage, &Router::new(router), vec![msg])?.add_event(event.into())
+        }
+        Some(token) => {
+            let event = AxelarnetGatewayEvent::ContractCalled {
+                msg: msg.clone(),
+                payload: call_contract.payload,
+                token: Some(token.clone()),
+            };
+
+            Response::new()
+                .add_message(WasmMsg::Execute {
+                    contract_addr: nexus_gateway.to_string(),
+                    msg: to_json_binary(&nexus_gateway::msg::ExecuteMsg::RouteMessageWithToken(
+                        msg,
+                    ))
+                    .expect("failed to serialize route message with token"),
+                    funds: vec![token],
+                })
```

Comment:
don't we have a client yet?

Reply:
Nope. I don't mind doing it in this same PR, but it's already pretty big.

Reply:
we can do a follow-up